### PR TITLE
Summarizing chat memory

### PR DESCRIPTION
**Pull request by PaperFly-web**
## Change
Added `SummarizingChatMemory` class to support automatic summarization of chat history.

Features include:
- Automatic summarization when message count exceeds a configurable threshold
- Support for both static and dynamic `maxMessages` and `summarizeThreshold`
- Dynamic system prompt generation per memory ID for context-aware summaries
- Automatic cleanup of orphan `ToolExecutionResultMessage` when associated `AiMessage` is evicted
- Ensures `SystemMessage` is always preserved at the head of memory
- Works with any `ChatMemoryStore` for persistence
- Builder API with clear and flexible configuration
- Comprehensive unit tests covering edge cases and dynamic configuration

Example usage:

```java
ChatMemory chatMemory = SummarizingChatMemory.builder()
    .chatModel(chatModel)
    .maxMessages(4)
    .summarizeThreshold(2)
    .dynamicSystemPrompt(id -> "Dynamic prompt for memory " + id)
    .build();
```

## General checklist

- [X] There are no breaking changes (API, behaviour)
- [X] I have added unit and/or integration tests for my change
- [X] The tests cover both positive and negative cases
- [X] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [X] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green

- [ ] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)

## Checklist for adding new maven module

- [ ] I have added my new module in the root `pom.xml` and `langchain4j-bom/pom.xml`

## Checklist for adding new embedding store integration

- [ ] I have added a `{NameOfIntegration}EmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreRemovalIT` that extends from `EmbeddingStoreWithRemovalIT`

## Checklist for changing existing embedding store integration

- [ ] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

### CODE REVIEW

**Comment by dliubarskyi** [comment]
cc @cescoffier you might be interested in this PR

File: langchain4j/src/main/java/dev/langchain4j/memory/chat/SummarizingChatMemory.java [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+        }
+
+        int insertIndex = (!messages.isEmpty() && messages.get(0) instanceof SystemMessage) ? 1 : 0;
+        messages.add(insertIndex, generateSummary(removedMessages));
+    }
+
+    /**
+     * Generate a concise AI summary for the removed messages.
+     */
+    private AiMessage generateSummary(List<ChatMessage> messages) {
```
Comment (dliubarskyi) [inline langchain4j/src/main/java/dev/langchain4j/memory/chat/SummarizingChatMemory.java]: Not sure that summary should reside in the `AiMessage`. Perhaps better to keep it in the `UserMessage`?

File: langchain4j/src/test/java/dev/langchain4j/memory/chat/SummarizingChatMemoryTest.java [hunk 2]
```diff
@@ -0,0 +1,782 @@
...
+import dev.langchain4j.model.openai.OpenAiChatModel;
+import java.util.List;
+import java.util.stream.Stream;
+import org.assertj.core.api.WithAssertions;
+import org.junit.jupiter.api.condition.EnabledIfEnvironmentVariable;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+@EnabledIfEnvironmentVariable(named = "deepseek-key", matches = ".+")
```
Comment (dliubarskyi) [inline langchain4j/src/test/java/dev/langchain4j/memory/chat/SummarizingChatMemoryTest.java]:
```suggestion
@EnabledIfEnvironmentVariable(named = "OPENAI_API_KEY", matches = ".+")
```

File: langchain4j/src/test/java/dev/langchain4j/memory/chat/SummarizingChatMemoryTest.java [hunk 3]
```diff
@@ -0,0 +1,782 @@
...
+                .build();
+
+        // Add the first three messages
+        chatMemory.add(userMessage("hello"));
+        chatMemory.add(userMessage("world"));
+        chatMemory.add(userMessage("I am lisi"));
+
+        // Verify that the first three messages are still retained
+        assertThat(chatMemory.messages())
+                .usingRecursiveFieldByFieldElementComparatorIgnoringFields("id", "timestamp", "createdAt")
```
Comment (dliubarskyi) [inline langchain4j/src/test/java/dev/langchain4j/memory/chat/SummarizingChatMemoryTest.java]: why is this required?

File: langchain4j/src/test/java/dev/langchain4j/memory/chat/SummarizingChatMemoryTest.java [hunk 4]
```diff
@@ -0,0 +1,782 @@
...
+                .containsExactly(userMessage("hello"), userMessage("world"), userMessage("I am lisi"));
+
+        // Add the fourth message, which may trigger summarization logic
```
Comment (dliubarskyi) [inline langchain4j/src/test/java/dev/langchain4j/memory/chat/SummarizingChatMemoryTest.java]: why this comment says that it "may" trigger? This test should be fully deterministic

File: langchain4j/src/main/java/dev/langchain4j/memory/chat/SummarizingChatMemory.java [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+        SystemMessage systemPrompt = SystemMessage.from(systemPromptFunction.apply(id));
+        messages.add(0, systemPrompt);
```
Comment (dliubarskyi) [inline langchain4j/src/main/java/dev/langchain4j/memory/chat/SummarizingChatMemory.java]: this seems odd, why do you insert the summarization message in the memory?

Comment (dliubarskyi) [reply langchain4j/src/main/java/dev/langchain4j/memory/chat/SummarizingChatMemory.java]: my bad - it is not memory, but removed messages. The name of the variable is confusing. In any case it is not the best practice to modify collections passed into method as arguments

File: langchain4j/src/main/java/dev/langchain4j/memory/chat/SummarizingChatMemory.java [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+import dev.langchain4j.store.memory.chat.ChatMemoryStore;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+
+/**
+ * A chat memory implementation that automatically summarizes conversation history
+ * when the number of messages exceeds a configurable threshold.
```
Comment (dliubarskyi) [inline langchain4j/src/main/java/dev/langchain4j/memory/chat/SummarizingChatMemory.java]:
```suggestion
 * when the number of messages exceeds a configurable `maxMessages`.
```
Otherwise it can be confused with `summarizeThreshold`

Comment (dliubarskyi) [reply langchain4j/src/main/java/dev/langchain4j/memory/chat/SummarizingChatMemory.java]: But in general I think it would be more useful to have `maxTokens` setting instead of `maxMessages`, as it is much more precise and controllable

File: langchain4j/src/main/java/dev/langchain4j/memory/chat/SummarizingChatMemory.java [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+ *
+ * <p>This class behaves similarly to {@link MessageWindowChatMemory} but adds automatic
+ * summarization capabilities using an underlying {@link ChatModel}.
+ *
+ * <p>Key features:
+ * <ul>
+ *     <li>Supports both static and dynamic configuration of <b>maxMessages</b>:
+ *         the maximum number of messages retained before triggering evictions.</li>
+ *     <li>Supports both static and dynamic configuration of <b>summarizeThreshold</b>:
```
Comment (dliubarskyi) [inline langchain4j/src/main/java/dev/langchain4j/memory/chat/SummarizingChatMemory.java]: I am not sure `summarizeThreshold` is the clear enough name, perhaps something like `maxMessagesToSummarize`?

File: langchain4j/src/main/java/dev/langchain4j/memory/chat/SummarizingChatMemory.java (hunk 1 above)
Comment (PaperFly-web) [reply langchain4j/src/main/java/dev/langchain4j/memory/chat/SummarizingChatMemory.java]: Yes, you're absolutely right — keeping the summary in the UserMessage makes more sense. I’ll update it to use UserMessage instead.

File: langchain4j/src/test/java/dev/langchain4j/memory/chat/SummarizingChatMemoryTest.java (hunk 2 above)
Comment (PaperFly-web) [reply langchain4j/src/test/java/dev/langchain4j/memory/chat/SummarizingChatMemoryTest.java]: OK, I will replace it.

File: langchain4j/src/test/java/dev/langchain4j/memory/chat/SummarizingChatMemoryTest.java (hunk 4 above)
Comment (PaperFly-web) [reply langchain4j/src/test/java/dev/langchain4j/memory/chat/SummarizingChatMemoryTest.java]: This is my mistake — I didn’t check this comment carefully. I’ll review it again to make sure it’s correct.This probably won’t be triggered now — it will only be triggered the next time something is added.

File: langchain4j/src/main/java/dev/langchain4j/memory/chat/SummarizingChatMemory.java (hunk 5 above)
Comment (PaperFly-web) [reply langchain4j/src/main/java/dev/langchain4j/memory/chat/SummarizingChatMemory.java]: Okay, create a new copy of the list to avoid modifying the passed parameters.

File: langchain4j/src/main/java/dev/langchain4j/memory/chat/SummarizingChatMemory.java (hunk 7 above)
Comment (PaperFly-web) [reply langchain4j/src/main/java/dev/langchain4j/memory/chat/SummarizingChatMemory.java]: Thank you for your suggestion. Indeed, "maxMessagesToSummarize" has a better semantic meaning, and I will replace it.

File: langchain4j/src/main/java/dev/langchain4j/memory/chat/SummarizingChatMemory.java (hunk 6 above)
Comment (PaperFly-web) [reply langchain4j/src/main/java/dev/langchain4j/memory/chat/SummarizingChatMemory.java]: Indeed, maxTokens would be better. It can prevent the summary from exceeding the token limit of the large model due to an excessive number of messages. I will switch to using tokens to summarize the conversation.

Comment (PaperFly-web) [reply langchain4j/src/main/java/dev/langchain4j/memory/chat/SummarizingChatMemory.java]: However, if we switch to using maxTokens, it becomes difficult to control whether the generated summary exceeds maxTokens. Therefore, I think we can consider providing an interface function, allowing users to summarize according to their own strategies if they have any.

**Comment by dliubarskyi** [comment]
BTW there was an old PR that tried to do a similar thing, might be worth checking it for ideas: https://github.com/langchain4j/langchain4j/pull/1351

**Comment by PaperFly-web** [comment]
> BTW there was an old PR that tried to do a similar thing, might be worth checking it for ideas: #1351

Thank you for your reminder. I checked it out and feel that his version is a bit messy. He generated system messages through summarization. Logically speaking, system messages should not be modified casually; generally, it's better to generate user messages. However, I still appreciate your reminder. I will carefully look at his code again to gain some inspiration.

**Comment by douglas-DS** [comment]
@PaperFly-web @dliubarskyi  I’m wondering whether using a _maxMessages_ threshold is the best approach here. Most AI frameworks I’ve seen apply summarization based on a token window, typically controlled by a _maxTokens_ setting. This ensures the memory is kept within the configured context size. Would it make more sense to follow that pattern here as well?

**Comment by dliubarskyi** [comment]
@douglas-DS [agreed](https://github.com/langchain4j/langchain4j/pull/4003#discussion_r2519059833)

**Comment by PaperFly-web** [comment]
> @PaperFly-web @dliubarskyi I’m wondering whether using a _maxMessages_ threshold is the best approach here. Most AI frameworks I’ve seen apply summarization based on a token window, typically controlled by a _maxTokens_ setting. This ensures the memory is kept within the configured context size. Would it make more sense to follow that pattern here as well?
@douglas-DS 
Sorry, I've been busy with work lately and forgot to reply. I've also tried using token calculation, but the final summary generated by the large language model cannot be controlled to a specific number of tokens. So, you might have defined summarizing once with more than 1000 tokens, but this summary might generate more than 1000 tokens of data. This would cause a conflict between the summary and the user-defined token threshold. Therefore, I don't have a better solution to this problem at the moment.

**Comment by dliubarskyi** [comment]
Thanks a lot for your PR! Unfortunately, we have to close it. See more details here: https://github.com/langchain4j/langchain4j/issues/5443#issuecomment-4707891571
